In [1]:
import pandas as pd
pd.set_option('display.max_columns', None)
pd.set_option('precision', 2)

In [2]:
# Reading the CSV File
df = pd.read_csv('2020_elections.csv')

In [3]:
df.head()

State Total Ballots Counted (Estimate)  \
0  United States                      158,835,004   
1        Alabama                        2,306,587   
2         Alaska                          367,000   
3        Arizona                        3,400,000   
4       Arkansas                        1,212,030   

  Vote for Highest Office (President) VEP Turnout Rate  \
0                                 NaN           66.40%   
1                           2,297,295           62.60%   
2                                 NaN           69.80%   
3                                 NaN           65.50%   
4                           1,206,697           55.50%   

  Voting-Eligible Population (VEP) Voting-Age Population (VAP) % Non-citizen  \
0                      239,247,182                 257,605,088         7.80%   
1                        3,683,055                   3,837,540         2.30%   
2                          525,568                     551,117         3.40%   
3                        5,189,000                   5,798,473         8.90%   
4                        2,182,375                   2,331,171         3.60%   

      Prison  Probation   Parole Total Ineligible Felon Overseas Eligible  \
0  1,461,074  1,962,811  616,440              3,294,457         4,971,025   
1     25,898     50,997   10,266                 67,782               NaN   
2      4,293      2,074    1,348                  6,927               NaN   
3     38,520     76,844    7,536                 93,699               NaN   
4     17,510     36,719   24,698                 64,974               NaN   

  State Abv  
0       NaN  
1        AL  
2        AK  
3        AZ  
4        AR

Dropping rows and columns that are either irrelevant or unusable due to many null-value data entries

In [4]:
# Dropping United States row entry because the analysis is focusing ont he individual states, not the country as a whole.
df = df.drop([0], axis=0)

In [5]:
# Dropping Vote for Highest Office (President) and Overseas Eligible due to many null-value entries
df = df.drop(['Vote for Highest Office (President)'], axis=1)
df = df.drop(['Overseas Eligible'], axis=1)

In [6]:
df.head()

State Total Ballots Counted (Estimate) VEP Turnout Rate  \
1     Alabama                        2,306,587           62.60%   
2      Alaska                          367,000           69.80%   
3     Arizona                        3,400,000           65.50%   
4    Arkansas                        1,212,030           55.50%   
5  California                       16,800,000           64.70%   

  Voting-Eligible Population (VEP) Voting-Age Population (VAP) % Non-citizen  \
1                        3,683,055                   3,837,540         2.30%   
2                          525,568                     551,117         3.40%   
3                        5,189,000                   5,798,473         8.90%   
4                        2,182,375                   2,331,171         3.60%   
5                       25,962,648                  30,783,255        15.00%   

    Prison Probation   Parole Total Ineligible Felon State Abv  
1   25,898    50,997   10,266                 67,782        AL  
2    4,293     2,074    1,348                  6,927        AK  
3   38,520    76,844    7,536                 93,699        AZ  
4   17,510    36,719   24,698                 64,974        AR  
5  104,730         0  102,586                207,316        CA

The data entries are not integers or floats, so currently they are unusable; they need to be converted.

In [7]:
# To convert regular numbers and remove commas
def convert(column):
    lst = []
    for num in column:
        string = ''
        if len(num)<=7:
            string=num[-3:]+string
            string=num[-7:-4]+string
        else:
            string=num[-3:]+string
            string=num[-7:-4]+string
            string=num[:-8]+string
        lst.append(int(string))
    return lst
  
# To convert percents into numbers
def convertPercent(column):
    for i in range(0,len(column)):
      column[i] = float(column[i][:-1])
    return column

In [8]:
df['Total Ballots Counted (Estimate)'] = convert(list(df['Total Ballots Counted (Estimate)']))
df['Voting-Eligible Population (VEP)'] = convert(list(df['Voting-Eligible Population (VEP)']))
df['Voting-Age Population (VAP)'] = convert(list(df['Voting-Age Population (VAP)']))
df['Prison'] = convert(list(df['Prison']))
df['Probation'] = convert(list(df['Probation']))
df['Parole'] = convert(list(df['Parole']))
df['Total Ineligible Felon'] = convert(list(df['Total Ineligible Felon']))

df['VEP Turnout Rate'] = convertPercent(list(df['VEP Turnout Rate']))
df['% Non-citizen'] = convertPercent(list(df['% Non-citizen']))

In [9]:
df.head()

State  Total Ballots Counted (Estimate)  VEP Turnout Rate  \
1     Alabama                           2306587              62.6   
2      Alaska                            367000              69.8   
3     Arizona                           3400000              65.5   
4    Arkansas                           1212030              55.5   
5  California                          16800000              64.7   

   Voting-Eligible Population (VEP)  Voting-Age Population (VAP)  \
1                           3683055                      3837540   
2                            525568                       551117   
3                           5189000                      5798473   
4                           2182375                      2331171   
5                          25962648                     30783255   

   % Non-citizen  Prison  Probation  Parole  Total Ineligible Felon State Abv  
1            2.3   25898      50997   10266                   67782        AL  
2            3.4    4293       2074    1348                    6927        AK  
3            8.9   38520      76844    7536                   93699        AZ  
4            3.6   17510      36719   24698                   64974        AR  
5           15.0  104730          0  102586                  207316        CA

In [10]:
# Adding another column indicating which party each state voted for in the election
df['Party'] = ['R', 'R', 'D', 'R', 'D','D','D','D','D','R','D','D','R','D','R','R','R','R','R','D','D','D','D','D','R','R','R','R','D','D','D','D','D','R','R','R','R','D','D','D','R','R','R','R','R','D','D','D','R','D','R',]

In [11]:
df.head()

State  Total Ballots Counted (Estimate)  VEP Turnout Rate  \
1     Alabama                           2306587              62.6   
2      Alaska                            367000              69.8   
3     Arizona                           3400000              65.5   
4    Arkansas                           1212030              55.5   
5  California                          16800000              64.7   

   Voting-Eligible Population (VEP)  Voting-Age Population (VAP)  \
1                           3683055                      3837540   
2                            525568                       551117   
3                           5189000                      5798473   
4                           2182375                      2331171   
5                          25962648                     30783255   

   % Non-citizen  Prison  Probation  Parole  Total Ineligible Felon State Abv  \
1            2.3   25898      50997   10266                   67782        AL   
2            3.4    4293       2074    1348                    6927        AK   
3            8.9   38520      76844    7536                   93699        AZ   
4            3.6   17510      36719   24698                   64974        AR   
5           15.0  104730          0  102586                  207316        CA   

  Party  
1     R  
2     R  
3     D  
4     R  
5     D

What is the average percent of non-citizen residents of each state and the District?

In [12]:
average = df['% Non-citizen'].mean()

d_states = df['Party']=="D"
democrats = df[d_states]

r_states = df['Party']=="R"
republicans = df[r_states]

In [13]:
average

5.443137254901959

In [14]:
democrats['% Non-citizen'].mean()

6.907692307692306

In [15]:
republicans['% Non-citizen'].mean()

3.92

On average, each state and the district had 5.44% of its residents as non-citizens. Democratic-voting states had an average of 6.91% of their populations as non-citizens, and Republican-voting states had an average of 3.92% of their populations as non-citizens.

In [16]:
import plotly.express as px

In [17]:
# Moving rows in the dataset so that when the graphs are autogenerated the Democratic-voting states appear blue and the Republican-voting states appear red.
df = df.loc[3:5].append(df.loc[0:2]).append(df.loc[5:]).reset_index()
df = df.drop([2])
df = df.drop(['index'], axis = 1)
df

State  Total Ballots Counted (Estimate)  VEP Turnout Rate  \
0                Arizona                           3400000              65.5   
1               Arkansas                           1212030              55.5   
3                Alabama                           2306587              62.6   
4                 Alaska                            367000              69.8   
5             California                          16800000              64.7   
6               Colorado                           3295000              76.4   
7            Connecticut                           1850000              71.1   
8               Delaware                            507805              70.5   
9   District of Columbia                            350000              64.7   
10               Florida                          11150000              71.7   
11               Georgia                           5025000              68.1   
12                Hawaii                            579165              57.5   
13                 Idaho                            875000              67.7   
14              Illinois                           6100000              67.6   
15               Indiana                           3070000              61.4   
16                  Iowa                           1825000              78.6   
17                Kansas                           1340000              64.2   
18              Kentucky                           2150951              64.9   
19             Louisiana                           2155000              63.9   
20                 Maine                            860000              79.2   
21              Maryland                           3050000              70.7   
22         Massachusetts                           3725000              73.4   
23              Michigan                           5559186              73.6   
24             Minnesota                           3290000              79.9   
25           Mississippi                           1330000              60.4   
26              Missouri                           3050000              66.3   
27               Montana                            612055              73.1   
28              Nebraska                            948852              68.6   
29                Nevada                           1370000              63.6   
30         New Hampshire                            814092              75.4   
31            New Jersey                           4495000              73.0   
32            New Mexico                            923612              61.0   
33              New York                           8930000              65.3   
34        North Carolina                           5600000              72.2   
35          North Dakota                            365000              64.6   
36                  Ohio                           6100000              68.9   
37              Oklahoma                           1565000              55.0   
38                Oregon                           2404642              75.2   
39          Pennsylvania                           6900000              70.5   
40          Rhode Island                            520000              65.0   
41        South Carolina                           2532995              64.5   
42          South Dakota                            427529              66.0   
43             Tennessee                           3060000              59.7   
44                 Texas                          11300000              60.2   
45                  Utah                           1450000              66.2   
46               Vermont                            370000              74.0   
47              Virginia                           4425000              71.4   
48            Washington                           4100000              75.4   
49         West Virginia                            795000              57.0   
50             Wisc

In [18]:
averageBarChart = px.bar(df, x='State Abv', y='% Non-citizen', color='Party', title='Percent of Non-Citizen Residents by State (and District) and Party')
averageBarChart.show()

In [19]:
scatter = px.scatter(df, x="Voting-Eligible Population (VEP)", y="Total Ineligible Felon", hover_name='State Abv', 
              title='Voting Eligible Population vs Total Ineligible Population', color='Party', trendline='ols')
scatter.show()

/usr/local/lib/python3.6/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning:

pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.



How does the ratio of the Voting Eligible Population to the Voting Age Population differ between the red and blue states?

In [20]:
# Creating a new column in the dataset to contain the ratio
df['VEP/VAP'] = df['Voting-Eligible Population (VEP)']/df['Voting-Age Population (VAP)']
df.head()

State  Total Ballots Counted (Estimate)  VEP Turnout Rate  \
0     Arizona                           3400000              65.5   
1    Arkansas                           1212030              55.5   
3     Alabama                           2306587              62.6   
4      Alaska                            367000              69.8   
5  California                          16800000              64.7   

   Voting-Eligible Population (VEP)  Voting-Age Population (VAP)  \
0                           5189000                      5798473   
1                           2182375                      2331171   
3                           3683055                      3837540   
4                            525568                       551117   
5                          25962648                     30783255   

   % Non-citizen  Prison  Probation  Parole  Total Ineligible Felon State Abv  \
0            8.9   38520      76844    7536                   93699        AZ   
1            3.6   17510      36719   24698                   64974        AR   
3            2.3   25898      50997   10266                   67782        AL   
4            3.4    4293       2074    1348                    6927        AK   
5           15.0  104730          0  102586                  207316        CA   

  Party  VEP/VAP  
0     D     0.89  
1     R     0.94  
3     R     0.96  
4     R     0.95  
5     D     0.84

In [21]:
d_states = df['Party']=="D"
democrats = df[d_states]

r_states = df['Party']=="R"
republicans = df[r_states]

In [22]:
democrats['VEP/VAP'].mean()

0.9228606005459296

In [23]:
republicans['VEP/VAP'].mean()

0.9468045660854331

In [24]:
bar = px.bar(df, x='State Abv', y='VEP/VAP', color='Party', title='Ratio of Voting Eligible Population to Voting Age Population')
bar.show()

What is the average turnout rate in the states and District by party?

In [25]:
d_states = df['Party']=="D"
democrats = df[d_states]

r_states = df['Party']=="R"
republicans = df[r_states]

In [26]:
democrats['VEP Turnout Rate'].mean()

70.33846153846154

In [27]:
republicans['VEP Turnout Rate'].mean()

65.104

The average voter-eligible population turnout rate in Democratic-voting states was 70.34%, while the turnout rate in Republican-voting states was 65.10%.

In [28]:
turnout_bar = px.bar(df, x='State Abv', y='VEP Turnout Rate', color='Party', title='Voter Eligible Population Turnout Rate by Party (%)')
turnout_bar.show()

In [29]:
d_states = df['Party']=="D"
democrats = df[d_states]

r_states = df['Party']=="R"
republicans = df[r_states]

In [30]:
democrats['Voting-Eligible Population (VEP)'].mean()

5144103.5

In [31]:
republicans['Voting-Eligible Population (VEP)'].mean()

4035447.76

In [32]:
df['Voting-Eligible Population (VEP)'].mean()

4600644.803921568

The average size of the Voter-Eligible Population of all 50 states and the District is 4600645 people, while the average size for the Democratic-voting states is 5144106 and the average size for the Republican-voting states is 4035448 people.

In [33]:
vep_bar = px.bar(df, x='State Abv', y='Voting-Eligible Population (VEP)', color='Party', title='Voting Eligible Population by State')
vep_bar.show()